### PR TITLE
When dealing with dupe arguments, prefer leafifying if possible

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89896

See code comment for details. I also had to do some extra fixes:

* `run_functionalized_fw_and_collect_metadata` now is able to handle duplicated arguments
* `aot_wrapper_dedupe` now always returns boxed compiled functions
* `aot_wrapper_dedupe` is now applied to inference compiler along with autograd compiler (preexisting)

Fixes https://github.com/pytorch/torchdynamo/issues/1939
Fixes DebertaV2ForQuestionAnswering DebertaForMaskedLM DebertaForQuestionAnswering DebertaV2ForMaskedLM

Repro command:

```
python benchmarks/dynamo/huggingface.py --performance --float32 -dcuda --training --inductor --no-skip --dashboard --only DebertaForQuestionAnswering --cold_start_latency
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @zou3519 @Chillee @samdow @soumith @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -1066,25 +1074,145 @@ def format_guard_bug_msg(aot_config, expected):
...
+#
+#           def f(x, y):
+#               x.transpose_(0, 1)
+#               y.transpose_(0, 2)
+#
+#           x = torch.randn(2, 3, 4)
+#           f(x, x)
+#
+#       The ordering of the transposes inside f dictates whether or not
+#       you get [4, 2, 3] or [3, 4, 2].  This means that you cannot precompute
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1133]: In theory, we could probably track the ordering that metadata mutation happened to inputs inside of `run_functionalized_fw_and_collect_metadata`, and replay them in that order? But that feels even more complicated than the 2-step thing that you're doing here.

File: functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -1066,25 +1074,145 @@ def format_guard_bug_msg(aot_config, expected):
...
+#
+def aot_wrapper_dedupe(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *, compiler_fn):
+    # Get information about whether or not flat_fn mutates its arguments
+    # or not
+    try:
+        with enable_python_dispatcher():
+            fw_metadata, _out, _num_aliasing_metadata_outs = run_functionalized_fw_and_collect_metadata(
+                flat_fn
+            )(*flat_args)
+    except RuntimeError as e:
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1166]: I'm actually a bit confused about why this is erroring now, when it wasn't before this PR? Functionalization should still always run before decomposing to prims.

File: functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -1066,25 +1074,145 @@ def format_guard_bug_msg(aot_config, expected):
...
+        # need to remove a duplicate.
+        leaf_flat_args = []
+        args_set = set()
+        ok = True
+
+        for i, a in enumerate(flat_args):
+            if a not in args_set:
+                args_set.add(a)
+                leaf_flat_args.append(a)
+            elif fw_metadata.mutated_input_info[i] == MutationType.none:
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1190]: Right now, won't we fallback to strategy #2 sometimes when we don't have to? It looks like when you have dupe args, it always uses the first instance directly, and tries to detach() all later instances. If you have:
```
def f(x, y):
    y.mutate_(...)
```

Then we could use in theory use strategy #1 by passing in `f(x.detach(), x)`.

If you think it's not worth beefing up strategy #1 today, that seems fine to me. AKA the bad situation would be we use strategy #2 when we could have used strategy #1, and strategy #2 ends up assert failing due to dynamo not producing enough guards. But we want to fix dynamo's guards later anyway, IIUC.

File: functorch/_src/aot_autograd.py [hunk 4]
```diff
@@ -1133,54 +1260,58 @@ def remove_dupe_args(args):
...
 
     @wraps(flat_fn)
     def wrapped_flat_fn(*args):
         return flat_fn(*add_dupe_args(args))
 
-    compiled_fn = aot_dispatch_deduplicated_autograd(wrapped_flat_fn, deduped_flat_args, aot_config)
+    compiled_fn = compiler_fn(wrapped_flat_fn, deduped_flat_args, aot_config)
+
+    if not hasattr(compiled_fn, "_boxed_call"):
+        compiled_fn = make_boxed_func(compiled_fn)
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1272]: Hmm shouldn't there be a single place in the code where we always make this happen?

File: functorch/_src/aot_autograd.py [hunk 5]
```diff
@@ -1107,15 +1235,14 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Tensor], aot_config: AOTConfi
 
     seen_args = {}
     keep_arg_mask = []
-    dropped_args = False
     add_dupe_map = {}
     duped_arg_len = len(flat_args)
+    leafified_inputs = False
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1240]: This is never set / used anywhere?

File: functorch/_src/aot_autograd.py [hunk 6]
```diff
@@ -1133,54 +1260,58 @@ def remove_dupe_args(args):
...
+
+    if not hasattr(compiled_fn, "_boxed_call"):
+        compiled_fn = make_boxed_func(compiled_fn)
 
     @wraps(compiled_fn)
-    def wrapped_compiled_fn(*args):
-        return compiled_fn(*remove_dupe_args(args))
+    def wrapped_compiled_fn(args):
+        deduped_args = remove_dupe_args(args)
+        args.clear()
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1277]: Oh, so we may have already been passed in a boxed func, but now we're replacing the old argument list with a new (deduped) argument list, so we need to make sure to clear the old list. I feel like that's an easy thing to forget to do if we end up doing this multiple times (but maybe that's just the cost of all the complexity)

File: functorch/_src/aot_autograd.py [hunk 7]
```diff
@@ -1133,54 +1260,58 @@ def remove_dupe_args(args):
...
+        return compiled_fn(deduped_args)
+    wrapped_compiled_fn._boxed_call = True
+
+    # This can be uncommented when we properly guard for duplicates,
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1281]: Just to spell this out - is the idea that we **know** that this is a problem today, so we definitely want the loud error to happen by default. But once we add the guarding for duplicates in dynamo, then we'll be more confident that dupe args work default, and so we'll be comfortable with the debug checks not running by default?

File: functorch/_src/aot_autograd.py [hunk 8]
```diff
@@ -1133,54 +1260,58 @@ def remove_dupe_args(args):
...
+                aot_config,
+                f"{describe_input(i, aot_config)} would be a duplicate of "
+                f"{describe_input(add_dupe_map[i], aot_config)}"
+            )
+        # This is only an error if there is metadata mutation on both of
+        # the duped arguments; in this case, we need to know what order
+        # the metadata mutation applies in.  You'll get the correct result
+        # otherwise, because a graph that assumes distinct inputs works if
+        # you dupe the inputs (the gradient contributions from each input
+        # will get summed up appropriately.)
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1303]: I see. So this assert is too conservative today. Is that also true for the asserts on line 1293? (That in theory we only want them to trip the assert(s) if >= 2 of the duped args had metadata mutations)

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1133]: I think this is the correct long term solution, it allows us to avoid dupe specializing entirely.

File: functorch/_src/aot_autograd.py (hunk 2 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1166]: The prim error is because there is a torchinductor test that is feeding a direct call to prim directly to AOTAutograd (no decomp involved).

Comment (bdhirsh) [reply functorch/_src/aot_autograd.py:1166]: Why didn't that test fail previously 🤔 (if it was running functionalization directly on the prims before)

File: functorch/_src/aot_autograd.py (hunk 3 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1190]: Your proposal here does not work because you may have:

```
def f(x, y):
    y.unsqueeze_(2)
    x.blah()
```

and if x is y, then the metadata mutation must reflect to x. Your alternate detach strategy still does not propagate the metadata.

File: functorch/_src/aot_autograd.py (hunk 4 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1272]: I think the correct fix is to have aot_dispatch_autograd natively support boxed calling convention, then all the wrappers can assume they are only dealing with boxed funcs. However, right now aot_dispatch_autograd does not return boxed func.

File: functorch/_src/aot_autograd.py (hunk 6 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1277]: There is no away around this with the boxed calling convention. You must clear the arg list as soon as you are done with it.

File: functorch/_src/aot_autograd.py (hunk 3 above)
Comment (bdhirsh) [reply functorch/_src/aot_autograd.py:1190]: whoops. Ok, different direction: If we're given:
```
def f(x, y):
    x.unsqueeze_(2)
    y.blah()
...
f(x, x)
```

Then we want to skip strategy 1 here, but won't the code above do strategy 1? Right now it just checks if the dupe arg has any mutations, but we also need to check if the original dupe (that was already added to `args_set`) also has any mutations.

File: functorch/_src/aot_autograd.py (hunk 7 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1281]: yup

File: functorch/_src/aot_autograd.py (hunk 2 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1166]: It's because the test doesn't require grad, so it goes the inference route for aot autograd, skipping functionalization

**Review by bdhirsh — APPROVED** [review]
I think this comment is still valid: https://github.com/pytorch/pytorch/pull/89896/files#r1036224447 but pre-approving!

File: functorch/_src/aot_autograd.py (hunk 3 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1190]: Yes, you're right!

Comment (ezyang) [reply functorch/_src/aot_autograd.py:1190]: Actually, it is fine! The reason is that when we run mutation detection, we do so mimicking the aliasing we were called with. So x is y when x.unsqueeze_(2) is called, and then in the end we say that BOTH x and y had their metadata mutated (even though there was only one metadata mutation call.) I wrote a test for this though which I'll commit in.
